homeworks/nagynorbert/5_oop/car_fleet_mpr_classes.py

```diff
@@ -0,0 +1,83 @@
+# Homework 5. Nagy Norbert
+
+class OutOfFuelError(Exception):
+    """Fuel is not enough for the defined distance."""
+    pass
+
+class TooMuchFuelError(Exception):
+    """Fuel is too much."""
+    pass
+
+class CarDefineError(Exception):
+    """Number of parameters for defining a car is not met with specification."""
+    pass
+
+class UserDefineError(Exception):
+    """Number of how many cars should manipulate is incorrect."""
+    pass
+
+class Car:
+    def __init__(self,brand:str,model:str,year:int):
+        self.brand = brand
+        self.model = model
+        self.year = year
+        self.mileage = 0
+        self.fuel_level = 100
+    
+    def __str__(self):
+        return f"Brand:{self.brand} -- Model:{self.model} -- Year:{self.year} -- Mileage:{self.mileage} -- Fuel_level:{self.fuel_level}"
+
+    def drive(self,km:float):
+        fuel_consumption = 0.1 * km
+        if fuel_consumption > self.fuel_level:
+            raise OutOfFuelError("Fuel is not enough in car.")
+        self.mileage += km
+        print(f"Mileage now: {self.mileage}")
+        self.fuel_level -= fuel_consumption
+        return True
+
+    def refuel(self,fuel_amount:float):
+        fuel_can_add = self.fuel_level + fuel_amount
+        if fuel_can_add > 100:
+            raise TooMuchFuelError("The amount of refuel is too much.")
+        self.fuel_level += fuel_amount
+        return True
+    
+class Fleet:
+    cars_in_fleet = []
+    number_of_cars = 0
+
+    def __init__(self,name:str):
+        self.name = name
+
+    def __str__(self):
+        return f"{self.name}"
+
+    def add_car_to_fleet(self,car:Car):
+        self.cars_in_fleet.append(car)
+        self.number_of_cars += 1 
+
+    def cars_equal(self,car1:Car,car2:Car):
+        if car1.brand == car2.brand and car1.model == car2.model and car1.year == car2.year:
+            return True
+        return False
+
+    def remove_car_from_fleet(self,car_remove:Car):
+        for idx,car in enumerate(self.cars_in_fleet):
```

Comment:
egyszerűbb ha csak megvizsgálod hogy a car része e a flottának:
```
if car in self.cars_in_fleet:
    self.cars_in_fleet.remove(car)
```
Ezzel a megoldással az __eq__()-t sem kell egyébként felülírni, mert teljesen jó ha memory address alapján akarja törölni az objektumot. Ha van két ugyanolyan gyártójú, típusú és évjáratú autó, akkor a memory address a legmegfelelőbb hogy azonosítsa a törléshez az objektumot, máskülönben nem biztos h az törlődne amit te szeretnél. Ha lenne egy rendszám is, ami egyedileg azonosítja, akkor az is lehetne az __eq__() alapja az összehasonlításhoz.

Reply:
Itt voltam bajban. Mivel én user input-ról kértem be a brand, model, year hármast és ebből példányosítottam a remove-hoz egy car objectet nem találtam soha a listámban. Megoldásként a Car class-ban felülírtam az __eq__-t és akkor tekintettem két autót egyenlőnek, ha a brand, model, year egyezett - kvázi adatbázis kulcsként értelmezve. A fleet class remove részéből kiszerveztem a main részbe azt a vizsgálatot, hogy user által megadott brand, model, year hármas alapján egyenlő-e ez az autó a listában lévő valamelyik másik autóval és ha igen, akkor a lista adott elemét kitöröltem. Azt hiszem nehezítettem azzal a munkámat, hogy a userre bíztam a választás lehetőségét, hogy melyik autót akarja törölni. Viszont nekem ez így remek gyakorló volt az eddigiek felhasználására is.